### PR TITLE
GTC-3142 Implement TCL titiler endpoint

### PR DESCRIPTION
**Pull request by manukala6**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [x] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: [GTC-3142](https://gfw.atlassian.net/browse/GTC-3142)

## What is the new behavior?

- Adds titiler endpoint for TCL v1.11
-
-

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

- To be consistent with the other titiler PRs, the TCL titler uses the `/dynamic` path. However a dynamic TCL tile cache service already exists, and this would be replacing. I'd appreciate input on whether I should use the `/dynamic` path or something new like `/titiler`

## Other information

[GTC-3142]: https://gfw.atlassian.net/browse/GTC-3142?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/routes/titiler/umd_tree_cover_loss.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    tree_cover_loss = TreeCoverLoss(
+        start_year=start_year,
+        end_year=end_year,
+        render_type=render_type,
+        tree_cover_density_threshold=tree_cover_density_threshold,
+        zoom=zoom
+    )
+
+    with COGReader("s3://gfw-data-lake-staging/umd_tree_cover_density_2010/v1.6/raster/epsg-4326/cog/default.tif") as reader:
```
Comment (danscales) [inline app/routes/titiler/umd_tree_cover_loss.py]: If tree_cover_density_threshold is None (wasn't specified at all), then you can skip this read of the TCD tile, and so be more efficient.

Comment (danscales) [reply app/routes/titiler/umd_tree_cover_loss.py]: Also, as far as I know, you need to make this be umd_tree_cover_density_2000/v1.8, since TCL goes back to 2000. 

Before checking in, you will need to change over to using GLOBALS.tree_cover_loss_filters from globals.py to set the versions of the dataset.  (See use in other files of filter_datasets and filter_dataset).  Look like GLOBAL.tree_cover_loss_filters can be the same as what I used for drivers, GLOBALS.tree_cover_loss_drivers.filter.

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    def create_mask(self):
+        mask = ~self.no_data
+
+        # TCL goes from 2001 to 2023 by default, only filter if start_year or end_year is specified
+        if self.start_year != 2001:
+            start_mask = self.tree_cover_loss_data >= (self.start_year - 2000)
+            mask &= start_mask
+
+        if self.end_year != 2023:
+            end_mask = self.tree_cover_loss_data <= (self.start_year - 2000)
```
Comment (jterry64) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: I think you meant `(self.end_year - 2000)` here

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        return alpha
+    
+    def create_true_color_rgb(self):
+        scale_pow = self.scale_intensity(self.zoom)
+        scaled_intensity = scale_pow(self.intensity).astype("uint8")
+
+        r = np.full(self.tree_cover_loss_data.shape, 228, dtype="float32")
+        g = (
+            np.ones(self.tree_cover_loss_data.shape, dtype="float32") * 102
+            + (72 - self.zoom)
```
Comment (jterry64) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: I'm guess this was pulled from the decoder function, but it'd be great if could document why we use this formula to get the values (if anyone still knows) in comments, in case we want to change with it in the future

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        scale_pow = self.scale_intensity(self.zoom)
+        scaled_intensity = scale_pow(self.intensity).astype("uint8")
+
+        alpha = (scaled_intensity if self.zoom < 13 else self.intensity) * self.mask
+        return np.clip(alpha, 0, 255).astype("uint8")
+    
+    @staticmethod
+    def scale_intensity(zoom) -> Callable:
+        """Power scale function for intensity scaling."""
+        exp = 0.3 + ((zoom - 3) / 20) if zoom < 11 else 1
```
Comment (jterry64) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: Same here, anyone know why why exactly we do this? It may also help with figuring out how to make the dynamic TCD filtering look better

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        domain = (0, 255)
+        scale_range = (0, 255)
+        m = scale_range[1] / domain[1] ** exp
```
Comment (jterry64) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: If we know what this represents, a more descriptive name than one letter would be useful for readability

**Review by jterry64 — CHANGES_REQUESTED** [review]
Looks good! For my comments on documentation, we might need to have a huddle where we try to reverse engineer what all those formulas are doing, so feel free to just merge up if you don't have that knowledge yourself.

File: app/routes/titiler/algorithms/tree_cover_loss.py (hunk 3 above)
Comment (manukala6) [reply app/routes/titiler/algorithms/tree_cover_loss.py]: @jterry64 We use RGB(228, 102, 153) for the "pink pixel", scaling down GB values with zoom level. I pulled this from the front-end decoder function, I'm assuming these numbers come from some trial-and-error in the past.

File: app/routes/titiler/algorithms/tree_cover_loss.py (hunk 4 above)
Comment (manukala6) [reply app/routes/titiler/algorithms/tree_cover_loss.py]: @jterry64 Added some comments to explain each step. We can huddle to discuss adapting this for filtering

File: app/routes/titiler/algorithms/tree_cover_loss.py (hunk 5 above)
Comment (manukala6) [reply app/routes/titiler/algorithms/tree_cover_loss.py]: @jterry64 Modified the variable name to `scaling_factor`

File: app/settings/globals.py [hunk 6]
```diff
@@ -17,6 +17,9 @@
     "tree_cover_density": {"dataset": "umd_tree_cover_density_2010", "version": "v1.6"},
 }
 
+tree_cover_loss_filters = {
+    "tree_cover_density": {"dataset": "umd_tree_cover_density_2010", "version": "v1.6"},
+}
```
Comment (danscales) [inline app/settings/globals.py:35]: ****As I mentioned in my previous comments, the tree cover density filter for TCL needs to be tcd_2000, right?  If not, maybe you can reply here and comment on why?  As I understand it, it should be tcd2000 since TCL starts in 2001.  Dist alerts uses tcd2010, because dist alerts didn't start until 2023.

Also, you really should merge completely up to present changes, since I have already defined a tree_cover_loss_filters value, and you are going to conflict on that.  (Unless you make the test exactly the same and using TCD 2000.)

File: app/routes/titiler/umd_tree_cover_loss.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    start_year: Optional[int] = Query(2001, ge=2001, le=2022, description="Only show loss for given year and after"),
+    end_year: Optional[int] = Query(2023, ge=2002, le=2023, description="Only show loss until given year."),
+    style: RenderType = Query(RenderType.encoded, description="Render true color or encoded tiles"),
+    tcd: Optional[int] = Query(None, ge=0, le=100)
+) -> Response:
+    """UMD Tree Cover Loss raster tiles."""
+
+    tile_x, tile_y, zoom = xyz
+    bands = ["default", "intensity"]
+    folder: str = f"s3://{DATA_LAKE_BUCKET}/{DATASET}/v1.11/raster/epsg-4326/cog"
```
Comment (solomon-negusse) [inline app/routes/titiler/umd_tree_cover_loss.py]: Version shouldn't be hard-coded here.

File: app/routes/titiler/umd_tree_cover_loss.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    tags=["Raster Tiles"],
+    response_description="PNG Raster Tile",
+)
+async def umd_tree_cover_loss_raster_tile(
+    *,
+    version: UmdTreeCoverLossVersions,
+    xyz: Tuple[int, int, int] = Depends(raster_xyz),
+    start_year: Optional[int] = Query(2001, ge=2001, le=2022, description="Only show loss for given year and after"),
+    end_year: Optional[int] = Query(2023, ge=2002, le=2023, description="Only show loss until given year."),
+    style: RenderType = Query(RenderType.encoded, description="Render true color or encoded tiles"),
```
Comment (solomon-negusse) [inline app/routes/titiler/umd_tree_cover_loss.py]: We are naming the field "render_type" in at least two other places and we should be consistent.

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        if self.start_year != 2001:
+            start_mask = self.tree_cover_loss_data >= (self.start_year - 2000)
+            mask &= start_mask
+
+        if self.end_year != 2023:
+            end_mask = self.tree_cover_loss_data <= (self.end_year - 2000)
+            mask &= end_mask
+
+        # Threshold by TCD if specified
+        if self.tree_cover_density_data is not None:
```
Comment (solomon-negusse) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: `if self.tree_cover_density_data is not None and self.tree_cover_density_threshold is not None`
will be cleaner and reduce the nesting

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            if self.tree_cover_density_threshold is not None:
+                density_mask = self.tree_cover_density_data.array[0, :, :] >= self.tree_cover_density_threshold
```
Comment (solomon-negusse) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: I see you're handling the scenario when all tcd is no data differently from how @danscales did it for carbon data (https://github.com/wri/gfw-tile-cache/blob/ee5f8edb47c9441fee8d90d2b9b54eaa154b40e3/app/routes/titiler/algorithms/tree_cover_loss_drivers.py#L71). Should we settle on a consistent behavior?

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+                mask &= density_mask
+
+        return mask
+
+    def create_encoded_rgb(self):
+        # Red = intensity
+        r = np.clip(self.intensity, 0, 255).astype("uint8")
```
Comment (solomon-negusse) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: nice one

File: app/routes/titiler/algorithms/tree_cover_loss.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        # Clip values to [0, 255] and convert to uint8
+        # Adapted from bugfix: https://gfw.atlassian.net/browse/GTC-2916 
+        g = np.clip(g, 0, 255).astype("uint8")
+        b = np.clip(b, 0, 255).astype("uint8")
+
+        return np.stack([r.astype("uint8"), g, b], axis=0)
+
+    def create_true_color_alpha(self):
+        scale_pow = self.scale_intensity(self.zoom)
+        scaled_intensity = scale_pow(self.intensity).astype("uint8")
```
Comment (solomon-negusse) [inline app/routes/titiler/algorithms/tree_cover_loss.py]: To not introduce avoidable latency at high zoom level, I suggest doing this only for the zoom levels that need it, 
Also, the scale function doesn't do anything for zoom < 11 (ln 138), so you can adjust the line below to match that (if 11 is correct).

**Review by solomon-negusse — CHANGES_REQUESTED** [review]
Hi @manukala6, nice to see this come out. I have left you comments for suggested changes.

File: app/settings/globals.py (hunk 6 above)
Comment (manukala6) [reply app/settings/globals.py:35]: Switched to TCD_2000 v1.8

File: app/routes/titiler/tree_cover_loss_core.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+async def tree_cover_loss_core(
+    *,
+    dataset: str,
+    version: str,
+    algorithm: Type[BaseAlgorithm],
+    xyz: Tuple[int, int, int],
+    start_year: Optional[int],
+    end_year: Optional[int],
+    render_type: RenderType,
+    tcd: Optional[TreeCoverDensityThreshold] = Query(
```
Comment (danscales) [inline app/routes/titiler/tree_cover_loss_core.py]: **This type should be TCLTreeCoverLossDensityThreshold

**Review by jterry64 — APPROVED** [review]
Looks good to me, the only thing I'm not a big fan of is having a `tree_cover_loss_core` file with just a common function listed under `routes`, because usually you expect a routes file. I think I'd prefer you just consolidate all three of these into one files, with the common function and two routes. But I'm fine with just merging this and we can address that later when we refactor the TiTiler stuff more generally.

File: app/routes/titiler/algorithms/tree_cover_loss_base.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+    def create_true_color_alpha(self):
+        # Scale intensity if zoom level < 11, otherwise use original intensity
+        if self.zoom < 11:
+            scale_pow = self.scale_intensity(self.zoom)
+            scaled_intensity = scale_pow(self.intensity).astype("uint8")
+        else:
+            scaled_intensity = self.intensity.astype("uint8")
+
+        alpha = (scaled_intensity if self.zoom < 13 else self.intensity) * self.mask
```
Comment (danscales) [inline app/routes/titiler/algorithms/tree_cover_loss_base.py:98]: ** Justin @-mentioned you on his PR yesterday that you would want to take some changes that he made to his PR.  The only one that changes code behavior is - change this line to:

       alpha = scaled_intensity * self.mask
 
just like what he has now in tree_cover_loss_drivers.py. (I had asked about why check on zoom level 13 was here.)

**Review by danscales — APPROVED** [review]
Looks good, except for the one fix I gave.
